packages/http/src/mocker/callback/callbacks.ts

```diff
@@ -0,0 +1,82 @@
+import { IHttpCallbackOperation, IHttpOperationRequest } from '@stoplight/types';
+import { resolveRuntimeExpressions } from '../../util/runtimeExpression';
+import { IHttpRequest, IHttpResponse } from '../../types';
+import fetch  from 'node-fetch';
+import * as Option from 'fp-ts/lib/Option';
+import * as Either from 'fp-ts/lib/Either';
+import { map, reduce } from 'fp-ts/lib/Array';
+import * as Reader from 'fp-ts/lib/Reader';
+import * as Task from 'fp-ts/lib/Task';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { head } from 'fp-ts/lib/Array';
+import { pipe } from 'fp-ts/lib/pipeable';
+import { generate as generateHttpParam } from '../generator/HttpParamGenerator';
+import { validateOutput } from '../../validator';
+import { parseResponse } from '../../util/response';
+import withLogger from '../../withLogger';
+import { violationLogger } from '../../util/logger';
+import { Logger } from 'pino';
+
+export function runCallback({ callback, request, response }: { callback: IHttpCallbackOperation, request: IHttpRequest, response: IHttpResponse }): Reader.Reader<Logger, Task.Task<void>> {
+  return withLogger(logger => {
+    const { url, ...requestInit } = assembleRequest({ resource: callback, request, response });
+    const logViolation = violationLogger(logger);
+
+    logger.info({ name: 'CALLBACK' }, `${callback.callbackName}: Making request to ${url}...`);
+
+    return pipe(
+      TaskEither.tryCatch(() => fetch(url, requestInit), Either.toError),
+      TaskEither.chain(parseResponse),
+      TaskEither.map(element => validateOutput({ resource: callback, element })),
+      TaskEither.map(violations => {
+        logger.info({ name: 'CALLBACK' }, `${callback.callbackName}: Request finished`);
+
+        pipe(
+          violations,
+          map(logViolation)
+        );
+      }),
+      TaskEither.fold(
+        error => {
+          logger.error({ name: 'CALLBACK' }, `${callback.callbackName}: Request failed: ${error.message}`);
+          return async () => undefined;
+        },
+        () => async () => undefined
+      ),
+    );
+  });
+}
+
+function assembleRequest({ resource, request, response }: { resource: IHttpCallbackOperation, request: IHttpRequest, response: IHttpResponse }) {
+  return {
+    url: resolveRuntimeExpressions(resource.path, request, response),
+    headers: Option.toUndefined(assembleHeaders(resource.request)),
+    body: Option.toUndefined(assembleBody(resource.request)),
+    method: resource.method,
+  };
+}
+
+function assembleBody(request?: IHttpOperationRequest): Option.Option<string> {
+  return pipe(
+    Option.fromNullable(request && request.body && request.body.contents),
+    Option.chain(head),
+    Option.chain(generateHttpParam),
+    Option.map(JSON.stringify),
```

Comment:
This is tricky. The problem is that `JSON.stringify` can effectively throw an exception, specifically when fed with circular objects. We need to decide if we want to pretend this is never going to happen or not.

I suggest we try to handle this accordingly — and use + `Either.fromOption` + `Either.stringifyJSON` function to return an Either. (Also `Option.fromEither(Either.stringifyJSON(value))`)

Reply:
Improved 👍

Reply:
I checked out again this flow — I suggest to use the following that is a little bit more verbose/readable

```ts
function assembleBody(request?: IHttpOperationRequest): Option.Option<{ body: string; mediaType: string }> {
  return pipe(
    Option.fromNullable(request),
    Option.mapNullable(request => request.body),
    Option.mapNullable(body => body.contents),
    Option.chain(head),
    Option.chain(param =>
      pipe(
        param,
        generateHttpParam,
        Option.map(body => ({ body, mediaType: param.mediaType }))
      )
    ),
    Option.chain(({ body, mediaType }) =>
      pipe(
        Either.stringifyJSON(body, () => undefined),
        Either.map(body => ({ body, mediaType })),
        Option.fromEither
      )
    )
  );
}
```

I've pushed this as part of #772